### PR TITLE
Configurable option to disable stamping of received header and Original Authentication Received header

### LINKED ISSUE #2850: Option to disable adding of Received and OAR Headers
Haraka | Haraka.js — Version: 2.8.24 and 2.8.25
-- | --
Node | v10.15.X and above
OS | Linux f3cd80ab-5be5-49a7-4711-d351 4.15.0-107-generic #108~16.04.1-Ubuntu SMP Fri Jun 12 02:57:13 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
openssl | OpenSSL 1.1.1  11 Sep 2018

Observed behavior:
1) Haraka adds the received header to the incoming email

image

2) If Authentication-Results header  found, Haraka adds the OAR.

image

Feature request:
1) If configuration file exist and values set to disable_received_header=true, then Haraka should skip adding of received header
https://github.com/haraka/Haraka/blob/4d521118a59ba897f7e0e5521671efd8546ec410/connection.js#L1540

2) If configuration file exist and values set to disable_auth_results_clean=true, then Haraka should skip adding of OAR.
https://github.com/haraka/Haraka/blob/4d521118a59ba897f7e0e5521671efd8546ec410/connection.js#L1645

Value Addition:
Provides the confidence that Incoming and received EML checksum are exactly matches when it processed through Haraka

### PR DESCRIPTION
**Pull request by adarshasp87**
Fixes #2850 

Changes proposed in this pull request:
- skip received header stamping if its disabled in configuration file
-  skip calling auth_results_clean method if its disabled in configuration file

### CODE REVIEW

**Review by msimerson — CHANGES_REQUESTED** [review]
Please use verb prefixed imperative names for config files, as you did with hide_received_header. When you have negative post-verb naming like `disable_auth_results_clean` we can end up with obtuse and harder to reason about statements like `if (!disable_auth_results_clean) {`.

Your first variable would be better named `clean_auth_results`.

I wonder why both options are needed. If you aren't adding a Received header, then you're essentially hiding the Haraka hop in the message headers. If you're doing that, it follows that you wouldn't clean and the OAR header. Am I missing something?

The other major issue with this setting is discoverability. There's no docs included here. The only way for someone to find this knob to twiddle is by reading the source. I'm also not a fan of just adding another entry to docs/CoreConfig.pm.

@core , what do you think of consolidating this and a few other settings to `smtp.ini`? IE, something like this:

```ini
diff --git a/config/smtp.ini b/config/smtp.ini
index 042d57db..e14e4bd0 100644
--- a/config/smtp.ini
+++ b/config/smtp.ini
@@ -43,3 +43,16 @@
 ;   after this time it will hard close. 30s is usually long enough to
 ;   wait for outbound connections to finish.
 ;force_shutdown_timeout=30
+
+[headers]
+add_received=true
+clean_original_auth_results=true
+
+; replace header_hide_version
+show_haraka_version=true
+
+; replace max_header_lines
+max_lines=1000
+
+; replace max_received_count
+max_received=100
```

File: connection.js [hunk 1]
```diff
@@ -1536,8 +1538,8 @@ class Connection {
             this.errors++;
             return this.respond(503, "RCPT required first");
         }
-
-        this.accumulate_data(`Received: ${this.received_line()}\r\n`);
+        if (!this.hide_received_header)
+            this.accumulate_data(`Received: ${this.received_line()}\r\n`);
```
Comment (msimerson) [inline connection.js:1542]: Either use brackets or make this a single line `if` statement.

File: connection.js [hunk 2]
```diff
@@ -1642,7 +1644,8 @@ class Connection {
             this.logwarn(`Incoming message reached maximum parsing limit of ${trans.MAX_HEADER_LINES} header lines`);
         }
 
-        this.auth_results_clean();   // rename old A-R headers
+        if (!this.disable_auth_results_clean)
+            this.auth_results_clean();   // rename old A-R headers
```
Comment (msimerson) [inline connection.js:1648]: same, brackets or single line.

Comment (adarshasp87) [reply connection.js]: Done

**Review by adarshasp87 — COMMENTED** [review]
changes are committed as per code review comments. 

@msimerson  consolidation of settings to smtp.ini is a very good idea.
Answer to your question of why both options are needed. We use Haraka SMTP to collect eml and not to route further. So it's important to get the eml in original form(to match the checksum when collected through Haraka)

**Comment by msimerson** [comment]
> So it's important to get the eml in original form.

I get that. But couldn't you use just a single setting to control both? When would you want to hide the Received header and not clean the OAR header?  Or vice versa?

I'm leaving this PR with my smtp.ini question open for a few days so @core members have a chance to chime in.

**Comment by adarshasp87** [comment]
> > So it's important to get the eml in original form.
> 
> I get that. But couldn't you use just a single setting to control both? When would you want to hide the Received header and not clean the OAR header? Or vice versa?
> We don't have such use cases. Introduced two separate settings not breaking the existing functionality or causing any problems.
> I'm leaving this PR with my smtp.ini question open for a few days so @core members have a chance to chime in.

**Comment by adarshasp87** [comment]
@msimerson Any update on this PR. We are eagerly looking for this feature.
If you think consolidation of settings to smtp.ini is a good option, I will spend sometime and raise another PR.

**Comment by msimerson** [comment]
If this PR instead added settings to smtp.ini, I'd have merged it already.

I was hoping another core member would comment. I really dislike the "just add another config file" pattern for settings. It's terrible for discovery (if the config file doesn't exist, the only way to discover the setting is reading the code or docs. Perusing the smtp.ini file is an obvious place for someone to look for and find the setting.

I also don't like `config.get('hide_received_header') === 'true';` comparing string values. That has a number of failure modes (especially in human expectations). We have robust boolean parsing already in the .ini file reader, which is another reason to put any boolean settings in ini files. If I merge this, I feel like I'm writing an IOU for future me to move this setting to smtp.ini, and then I have to also add backwards compatibility for it.

**Comment by adarshasp87** [comment]
> If this PR instead added settings to smtp.ini, I'd have merged it already.
> 
> I was hoping another core member would comment. I really dislike the "just add another config file" pattern for settings. It's terrible for discovery (if the config file doesn't exist, the only way to discover the setting is reading the code or docs. Perusing the smtp.ini file is an obvious place for someone to look for and find the setting.
> 
> I also don't like `config.get('hide_received_header') === 'true';` comparing string values. That has a number of failure modes (especially in human expectations). We have robust boolean parsing already in the .ini file reader, which is another reason to put any boolean settings in ini files. If I merge this, I feel like I'm writing an IOU for future me to move this setting to smtp.ini, and then I have to also add backwards compatibility for it.

**Comment by msimerson** [comment]
@adarshasp87 , have a look at #2855

**Comment by adarshasp87** [comment]
@msimerson Thanks you so much for your support. Eagerly Looking for this feature getting released in new Haraka version

**Comment by msimerson** [comment]
Go ahead and test drive #2855 for me.  Or wait until I get around to it..

**Comment by adarshasp87** [comment]
@msimerson Sure mate. I will post my findings soon

**Comment by msimerson** [comment]
@adarshasp87 , please post soon. I've been holding up a release pending your feedback. I'm planning to release Wednesday unless there's an issue.

**Comment by adarshasp87** [comment]
@msimerson My Apologies for the delayed response.  I have not observed any issues with the code.

**Pull request closed without merging**